### ISSUE TITLE
systemd ExecStop= Not Run

### ISSUE THREAD

**Issue by jbulchergist** [coreos/bugs#186]
I have unit files for various docker services, and I would like to run a docker container before they exit. This docker container manages DNS - there could be other conceivable uses for such "utility containers".

I have the appropriate `Requires=docker.service` and `After=docker.service` in the Unit file.

The `ExecStop` command is not run until after the docker.service stops. In fact, the first thing CoreOS seems to do when shutting down is to stop running docker containers. By the time systemd gets around to the unit file, the docker container is long gone - and in my case even the docker.service has already been shut down.

The implications of this issue are that **no docker Unit files ever gracefully shut down** as written, which seems like pretty significant flaw (even if it only affects a minority of users, since the containers themselves get shut down)..

My Unit files are below (a little worse for wear, as I have tried many different configurations)

```
registry-dns.service
-------------------
[Unit]
Description=Docker Registry DNS Registration
Before=registry.service registry-ui.service
After=network.target docker.service
Requires=network.target docker.service registry.service registry-ui.service
Conflicts=shutdown.target
[Service]
TimeoutStartSec=0
RemainAfterExit=yes
Type=oneshot
ExecStart=/usr/bin/docker run --rm argo:5000/nsupdate dns \
                add \
                --address 192.168.1.60 \
                --domain foo.local \
                --host argo \
                --nameserver 192.168.1.50 \
                --nameserver 192.168.1.51
ExecStop=/usr/bin/docker run --rm argo:5000/nsupdate dns \
                delete \
                --address 192.168.1.60 \
                --domain foo.local \
                --host argo \
                --nameserver 192.168.1.50 \
                --nameserver 192.168.1.51
[X-Fleet]
MachineOf=registry.service
[Install]
RequiredBy=registry.service registry-ui.service
Also=registry.service registry-ui.service
cat: registry-dns.service.requires: Is a directory
```

```
registry-ui.service
------------------
[Unit]
Description=Docker Registry UI
After=registry.service registry-dns.service
Requires=registry.service registry-dns.service
Conflicts=shutdown.target
[Service]
TimeoutStartSec=0
Restart=on-failure
ExecStart=/usr/bin/docker start -a RegistryUI
ExecStop=/usr/bin/docker stop RegistryUI

[X-Fleet]
MachineOf=registry.service

[Install]
RequiredBy=registry.service registry-dns.service
Also=registry.service registry-dns.service
cat: registry-ui.service.requires: Is a directory
```

```
registry.service
-----------------
[Unit]
Description=Docker Registry
Requires=docker.service registry-dns.service registry-ui.service
After=docker.service registry-dns.service
Before=registry-ui.service
Conflicts=shutdown.target reboot.target

[Service]
TimeoutStartSec=0
Restart=on-failure
ExecStart=/usr/bin/docker start -a Registry
ExecStop=/bin/echo deleting dns
ExecStop=/usr/bin/docker run --rm argo:5000/nsupdate dns delete --address 192.168.1.60 --domain foo.local --host argo --nameserver 192.168.1.50 --nameserver 192.168.1.51
ExecStop=/bin/echo stopping registry
ExecStop=/usr/bin/docker stop Registry
ExecStop=/bin/echo stopped registry

[X-Fleet]
MachineID=coreos1

[Install]
WantedBy=multi-user.target
RequiredBy=registry-ui.service registry-dns.service docker.service
Also=registry-ui.service registry-dns.service
cat: registry.service.requires: Is a directory
```

**Comment by philips** [comment]
Update on this, it is fixed in systemd upstream and I am patching Docker's libcontainer to fix it too: https://github.com/docker/libcontainer/pull/359

**Comment by mailsquad** [comment]
Great work, thanks!

**Comment by kreisys** [comment]
Awesome. Any idea when the changes will be released in coreos?

**Comment by vcaputo** [comment]
@kreisys fixed in the next alpha release

**Comment by crawford** [comment]
Fixed in [584.0.0](https://github.com/coreos/manifest/releases/tag/v584.0.0).
